### DISCUSSION TITLE
sdk link issue with libnvinfer.so rpath and libcudla.so

### DISCUSSION THREAD

**Discussion by aurelien-enchanted-tools** [OE4T/meta-tegra#1383] [Q&A]
Hello, I met an issue with the sdk and cmake during the link step 

poky: langdale
meta-tegra: langdale

```
ld: warning: libcudla.so.1, needed by /home/yoctouser/sdk/sysroots/armv8a-custom-linux/usr/lib/libnvinfer.so, not found (try using -rpath or -rpath-link)
ld: /home/yoctouser/sdk/sysroots/armv8a-custom-linux/usr/lib/libnvinfer.so: undefined reference to `cudlaSubmitTask@libcudla.so.1'
```

In the sdk, I have got:
- `/usr/local/oe-sdk-hardcoded-buildpath/sysroots/armv8a-custom-linux/usr/lib/libnvinfer.so` and
- `/usr/local/oe-sdk-hardcoded-buildpath/sysroots/armv8a-custom-linux/usr/local/cuda-11.4/lib/libcudla.so.1`

I check the c++ method cudlaSubmitTask is in `/usr/local/oe-sdk-hardcoded-buildpath/sysroots/armv8a-custom-linux/usr/local/cuda-11.4/lib/libcudla.so.1`

In the image:
```
TOOLCHAIN_HOST_TASK += " \
    nativesdk-packagegroup-cuda-sdk-host \
"
```

I would like to know:
- how to include the `/usr/local/oe-sdk-hardcoded-buildpath/sysroots/armv8a-custom-linux/usr/local/cuda-11.4/lib/` in the libnvinfer's RPATH, or better
- how to fix the issue directly in the tensorrt-core recipe which populates the libnvinfer.so

Thank you in advance.

**Comment by madisongh** [comment]
The necessary compiler and linker flags/options to make the CUDA libraries available should get added automatically by CMake if you've told it you're using CUDA, either by specifying `CUDA` in the `LANGUAGES` clause in your `project()` command, or through `enable_language(CUDA)` (or, if you're using a very old version of CMake, by using `FindCUDA()`).  The SDK sets up [environment variables](https://github.com/OE4T/meta-tegra/blob/langdale/recipes-devtools/cuda/cuda-environment/cuda_environment_setup.sh.in) and provides a [CMake "toolchain" file](https://github.com/OE4T/meta-tegra/blob/langdale/recipes-devtools/cuda/cuda-environment/cuda_toolchain.cmake) for this purpose.  But you have to tell CMake about CUDA being used in your project.

**Reply by aurelien-enchanted-tools** [reply]
langdale cmake version: 3.24.2

I added in my yocto image config
```
TOOLCHAIN_HOST_TASK += " \
   nativesdk-cuda-environment \
   nativesdk-packagegroup-cuda-sdk-host \
"
 
TOOLCHAIN_TARGET_TASK += "\
    cuda-target-environment \
"
```
and in my CMakeLists.txt build with the sdk
```
project(custom LANGUAGES C CXX CUDA)
find_package(CUDA REQUIRED)
```
and I obtain the same error as previous.

**Reply by madisongh** [reply]
Can you enable tracing (use `--trace-expand` on the `cmake` command) and see where the additions might be getting lost?  I do have folks with cmake projects that call on the TensorRT libraries that don't run into this issue, so I'm wondering if either there's something in your environment that's causing the CUDA additions to get bypassed, or there's something in your cmake files that might be overriding what the macros are inserting for you into the compilation and linking flags.
